### ISSUE TITLE
Ensure thread safety for parallel tuning

### ISSUE THREAD

**Issue by rustyconover** [keras-team/keras-tuner#301]
When running multiple parallel workers on the same machine as the oracle these errors occur when starting multiple workers.

Any ideas?  Thank you.

```
Traceback (most recent call last):
  File "/home/ec2-user/.local/lib/python3.7/site-packages/kerastuner/engine/oracle.py", line 323, in reload
    super(Oracle, self).reload(self._get_oracle_fname())
  File "/home/ec2-user/.local/lib/python3.7/site-packages/kerastuner/engine/stateful.py", line 64, in reload
    self.set_state(state)
  File "/home/ec2-user/.local/lib/python3.7/site-packages/kerastuner/tuners/hyperband.py", line 298, in set_state
    super(HyperbandOracle, self).set_state(state)
  File "/home/ec2-user/.local/lib/python3.7/site-packages/kerastuner/engine/oracle.py", line 287, in set_state
    for tuner_id, trial_id in state['ongoing_trials'].items()}
  File "/home/ec2-user/.local/lib/python3.7/site-packages/kerastuner/engine/oracle.py", line 287, in <dictcomp>
    for tuner_id, trial_id in state['ongoing_trials'].items()}
KeyError: '121381334a7c340f8281c6e2d78dbb7a'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "trainer.py", line 53, in <module>
    project_name='iso-demand2')
  File "/home/ec2-user/.local/lib/python3.7/site-packages/kerastuner/tuners/hyperband.py", line 380, in __init__
    **kwargs)
  File "/home/ec2-user/.local/lib/python3.7/site-packages/kerastuner/engine/multi_execution_tuner.py", line 58, in __init__
    oracle, hypermodel, **kwargs)
  File "/home/ec2-user/.local/lib/python3.7/site-packages/kerastuner/engine/tuner.py", line 103, in __init__
    overwrite=overwrite)
  File "/home/ec2-user/.local/lib/python3.7/site-packages/kerastuner/engine/base_tuner.py", line 72, in __init__
    self.directory, self.project_name, overwrite=overwrite)
  File "/home/ec2-user/.local/lib/python3.7/site-packages/kerastuner/engine/oracle.py", line 298, in _set_project_dir
    self.reload()
  File "/home/ec2-user/.local/lib/python3.7/site-packages/kerastuner/engine/oracle.py", line 329, in reload
    'project at: {}'.format(self._project_dir))
RuntimeError: Error reloading `Oracle` from existing project. If you did not mean to reload from an existing project, change the `pro
ject_name` or pass `overwrite=True` when creating the `Tuner`. Found existing project at: results_dir7/iso-demand2
```

**Comment by rustyconover** [comment]
This happens when starting the chief first, then starting the workers.

If I start the workers first then the chief the exception doesn't happen.

**Comment by ag-tcm** [comment]
I also got these exceptions under the same circumstances. However, it wasn't consistent. Sometimes I start a worker after the chief and other workers are already running and it starts fine, other times it quits with the same exceptions as stated above.

**Comment by winginitau** [comment]
Any update on this? In my case, very rarely I get this error on one worker which is attached over a slow connection (10MB LAN). Does not seem to have any dependency on start order. However, I have a unexpected long tuner search underway and decided to join a high-end lapop as a worker for some extra power. It is connected over WIFI and won't start at all. 

Perhaps a timeout in the keras code loading the oracle json file?

**Comment by hvavhp** [comment]
I suspect this is due to some race conditions between the chief and the worker when they're both accessing the same state directory. Here's the code where the error occur in file `oracle.py` (line 335):
```python
def set_state(self, state):
    # `self.trials` are saved in their own, Oracle-agnostic files.
    self.ongoing_trials = {
        tuner_id: self.trials[trial_id]
        for tuner_id, trial_id in state["ongoing_trials"].items()
    }
```
`trial_id` was apparently not in `self.trials`. Moving back from the trace, we see the root cause possibly here in `oracle.py` (line 371):
```python
def reload(self):
    # Reload trials from their own files.
    trial_fnames = tf.io.gfile.glob(
        os.path.join(self._project_dir, "trial_*", "trial.json")
    )
    for fname in trial_fnames:
        with tf.io.gfile.GFile(fname, "r") as f:
            trial_data = f.read()
        trial_state = json.loads(trial_data)
        trial = trial_lib.Trial.from_state(trial_state)
        self.trials[trial.trial_id] = trial
    try:
        super(Oracle, self).reload(self._get_oracle_fname())
    except KeyError:
        raise RuntimeError(
            "Error reloading `Oracle` from existing project. If you did not "
            "mean to reload from an existing project, change the `project_name` "
            "or pass `overwrite=True` when creating the `Tuner`. Found existing "
            "project at: {}".format(self._project_dir)
        )
```
In this method `self.trials` is populated with trials found in `self._project_dir`. Then `super(Oracle, self).reload(self._get_oracle_fname())` which does the following:
```python
def reload(self, fname):
    """Reloads this object using `set_state`.

    Args:
        fname: A string, the file name to restore from.
    """
    with tf.io.gfile.GFile(fname, "r") as f:
        state_data = f.read()
    state = json.loads(state_data)
    self.set_state(state)
```
Basically it reads a list of trials from `oracle.json`, and then tries to reconcile with `self.trials`. There's a tiny time window between listing the trials in `self._project_dir` and listing the trials in `oracle.json`. Both these places are probably populated by the chief. Normally this time window is so small that there shouldn't be a problem. However, when there are a lot of workers running, and trials are being computed very swiftly, new trials will be frequently generated in advance, and thus will lead to higher chances of coming across the following situation
1. trials are listed in `self._project_dir` in `oracle.json`
2. worker reads trials from `self._project_dir`
3. more trials are listed in `self._project_dir` in `oracle.json`
4. worker reads trials from `oracle.json`
5. worker reconciles between the two lists, and failed

Of course this situation could become far more popular if there's an excessive delay due to network communication, which makes step 2 above lasting arbitrarily long. Note that the above is entirely my own hypothesis, as I did not have the capacity to investigate further.

Thanks.

**Comment by haifeng-jin** [comment]
I cannot understand why this error would occur.
The `Oracle` would never be reloaded due to the following code:

https://github.com/keras-team/keras-tuner/blob/f6b88bf87a90fed970bb36f0242030dea0710a9d/keras_tuner/engine/base_tuner.py#L426-L436

`has_chief_oracle()` is always `True` in distributed setting, it just check if `KERASTUNER_ORACLE_IP` exists in env.
So the code above says: never save or reload the oracle if in a distributed setting.

However, if `Oracle` do get reloaded, the error makes sense.
The workers requested new `Trial`s from the chief `Oracle`.
The `tuner_id` of the workers get added to `Oracle.ongoing_trials`.
All these happend before the `Oracle` start to reload.
Then, when reloading, the `Oracle` uses the new `tuner_id`s in the `self.ongoing_trials` in the following code, which causes the `KeyError`.

https://github.com/keras-team/keras-tuner/blob/5e8ae50262b7316f292068483233e73501a45a22/keras_tuner/engine/oracle.py#L531-L536

The solution would be to hold the `Oracle` from generating new trials until its get fully reloaded.

**Comment by haifeng-jin** [comment]
@hvavhp Thank you so much! This gives me a really good understanding of the bug.

I would fix it in the codebase like this:
No worker should reload the oracle since there is only one real oracle in the chief. 
The oracle in chief would fully reload before starting the gRPC server to accept worker calls.

I will also remove the `reload()` call in `_set_project_dir()`, which is bug prone. We keep all the reloading of oracle to the `tuner.reload()`.

BTW, I am trying to polish the parallel tuning workflow recently. If you have any other feedbacks, please feel free to submit more issues or mention me in any old issue that is relevant.
Thank you!

**Comment by haifeng-jin** [comment]
This is resolved by #821.
